File: frontend/src/lib/actions/structure.ts
"use server";

export interface UnitPreview {
  name: string;
  code: string;
  credits: number;
  teaching_block: string;
  status: string;
  link: string;
}

export interface ProgrammePreview {
  programme_name: string;
  units: UnitPreview[];
}

export interface StructurePreviewResponse {
  results: ProgrammePreview[];
}

export async function getStructurePreview(link: string, years: string[]) {
  const yearNumbers = years.map((y) => parseInt(y.replace("Year ", "")));

  try {
    const response = await fetch(
      `${process.env.NEXT_PUBLIC_API_URL}/structure/preview`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          link: link,
          years: yearNumbers,
        }),
      },
    );

    if (!response.ok) throw new Error("Failed to fetch structure");

    const data: StructurePreviewResponse = await response.json();
    return data;
  } catch (error) {
    console.error("Scraper Error:", error);
    return null;
  }
}

File: backend/app/schemas/structure.py
from typing import List
from pydantic import BaseModel

class UnitPreview(BaseModel):
    name: str
    code: str
    credits: int
    teaching_block: str
    status: str
    link: str

class ProgrammePreview(BaseModel):
    programme_name: str
    start_year: int
    end_year: int
    units: List[UnitPreview]

class StructurePreviewResponse(BaseModel):
    results: List[ProgrammePreview]

class StructureCreateResponse(BaseModel):
    message: str

class PreviewPayload(BaseModel):
  link: str
  years: List[int]